src/Microsoft.Diagnostics.Monitoring.EventPipe/Configuration/LoggingSourceConfiguration.cs

```diff
@@ -5,51 +5,117 @@
 
         public override IList<EventPipeProvider> GetProviders()
         {
-            string filterSpec = _useAppFilters ? UseAppFilters : FormattableString.Invariant($"*:{_level:G}");
-
-            var providers = new List<EventPipeProvider>()
+            return new List<EventPipeProvider>()
             {
-
-                // Logging
                 new EventPipeProvider(
                     MicrosoftExtensionsLoggingProviderName,
-                    EventLevel.LogAlways,
-                    (long)(LoggingEventSource.Keywords.JsonMessage | LoggingEventSource.Keywords.FormattedMessage),
+                    _level,
+                    _keywords,
                     arguments: new Dictionary<string,string>
                         {
-
-                            { "FilterSpecs", filterSpec }
+                            { "FilterSpecs", _filterSpecs }
                         }
                 )
             };
+        }
+
+        private static string ToFilterSpecsString(IDictionary<string, LogLevel?> filterSpecs, bool useAppFilters)
+        {
+            if (!useAppFilters && (filterSpecs?.Count).GetValueOrDefault(0) == 0)
+            {
+                return String.Empty;
+            }
+
+            StringBuilder filterSpecsBuilder = new StringBuilder();
+
+            if (useAppFilters)
+            {
+                filterSpecsBuilder.Append("UseAppFilters");
+            }
+
+            if (null != filterSpecs)
+            {
+                foreach (KeyValuePair<string, LogLevel?> filterSpec in filterSpecs)
+                {
+                    if (!string.IsNullOrEmpty(filterSpec.Key))
+                    {
+                        if (filterSpecsBuilder.Length > 0)
+                        {
+                            filterSpecsBuilder.Append(";");
+                        }
+                        filterSpecsBuilder.Append(filterSpec.Key);
+                        if (filterSpec.Value.HasValue)
+                        {
+                            filterSpecsBuilder.Append(":");
+                            filterSpecsBuilder.Append(filterSpec.Value.Value.ToString("G"));
+                        }
+                    }
+                }
+            }
+
+            return filterSpecsBuilder.ToString();
+        }
 
-            return providers;
+        private static EventLevel ToEventLevel(LogLevel logLevel)
+        {
+            switch (logLevel)
+            {
+                case LogLevel.Trace:
+                    return EventLevel.LogAlways;
+                case LogLevel.Debug:
+                    return EventLevel.Verbose;
+                case LogLevel.Information:
+                    return EventLevel.Informational;
+                case LogLevel.Warning:
+                    return EventLevel.Warning;
+                case LogLevel.Error:
+                    return EventLevel.Error;
+                case LogLevel.Critical:
+                    return EventLevel.Critical;
```

Comment:
There is no corresponding `EventLevel` that would be equivalent to `LogLevel.None`, so I think throwing is acceptable here since it cannot be supported. I could special case `LogLevel.None` to throw NotSupportedException with a small explanation that it can be any other level that `LogLevel.None`.

In fact, I'm not certain that `LogLevel.Trace`/`EventLevel.LogAlways` is really supported based on what I see in https://github.com/dotnet/runtime/blob/1410c12ef8068b06518ec99b6cc259dd0c448892/src/libraries/Microsoft.Extensions.Logging.EventSource/src/LoggingEventSource.cs#L446. It might return `LogLevel.Critical` or `LogLevel.Debug` depending on the behavior of the `IsEnabled` method. Maybe I'll log a dotnet/runtime issue for this one.

Reply:
I specifically handle `LogLevel.None` to throw NotSupportedException and put a test in to make sure this doesn't change unexpectedly.